### ISSUE TITLE
[rush] ignores config options from common/config/rush/.npmrc

### ISSUE THREAD

**Issue by moltar** [microsoft/web-build-tools#1468]
## Please prefix the issue title with the project name i.e. [rush], [api-extractor] etc. ##

**Is this a feature or a bug?**

- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**

I've added `save-exact=true` to `common/config/rush/.npmrc` file, but that setting is being ignored.

The docs state:

```
# Rush uses this file to configure the package registry, regardless of whether the
# package manager is PNPM, NPM, or Yarn.  Prior to invoking the package manager,
```

So if it is only using `.npmrc` for registry info, then maybe it is a misnomer to name the file `.npmrc` as that makes it appear as if it will be able to configure other aspects of NPM.

Btw, I am aware there is `--exact` flag for `rush add`, but it is difficult to codify that in the team environment. That is why I usually use a local `.npmrc` file.

**If the issue is a bug, how can we reproduce it? Please provide detailed steps and include a GitHub branch if applicable. Your issue will get resolved faster if you can make it easy to investigate.**

1. Add `save-exact=true` to `common/config/rush/.npmrc`
2. Run `rush add $package_name`
3. Open `package.json` and observe that the version is not exact.

**What is the expected behavior?**

The version to be exact.

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** rush
* **Tool Version:** 5.11.2
* **Node Version:** 10.15.3
  * **Is this a LTS version?** yes
  * **Have you tested on a LTS version?** yes
* **OS:** macOS

**Comment by octogonz** [comment]
It might be better to model this as a setting in   `rush.json` instead of relying on `.npmrc`.  If I remember right, Yarn does not honor `.npmrc` settings, so `save-exact` is somewhat proprietary to NPM.  Also, the `.npmrc` file format is not very strictly validated, and it's designed with the idea that settings can come from many places (environment variables, command-line arguments, project-specific `.npmrc`, user-defined `.nprmc`).  This is very flexible, but e.g. in https://github.com/microsoft/web-build-tools/issues/1452 we found that this flexibility causes confusion, so that issue is proposing to reduce the influence of `.npmrc`.

Does this approach make sense to you?

**Comment by moltar** [comment]
Are you proposing to get rid of `.npmrc` dependency in Rush in favor of `rush.json`?

As in, all existing settings that Rush checks in `.npmrc` (which from the docs sounds like just the registry info).

Or are you proposing to just move `save-exact` into a `rush.json` setting?

**Comment by octogonz** [comment]
No, just to use it for settings that control the package manager, rather than settings that control Rush itself. For example, what if Yarn and NPM interpret the setting slightly differently; which semantics should Rush follow? Or what if there are special additional states for that setting; are we obligated to implement them all in `rush add`? Or what if we want to extend the setting with more features; will we extend the .npmrc schema?

**Comment by EvanLovely** [comment]
Any progress on this? When I use `rush add` I want `^` instead of `~`.

**Comment by pseudo-su** [comment]
I would also like to default `rush add` to use `^` instead of `~`. I like the idea of putting this as a config in `rush.json` because the existing/current `save-exact=true` isn't precise enough to cover all the different options for how you might want to set default version range strategy (eg `carat | tilde | exact`).

**Comment by ksjogo** [comment]
Yes, some config would be great.

**Comment by kg-currenxie** [comment]
Same here :))

**Comment by Js-Brecht** [comment]
Lately I've been wanting some control over the default semver style.  Should be pretty easy to implement, tbh.

@iclanton would it make sense to add a general "default cli options" property in `rush.json`
```
{
  "cliDefaults": {
    "semVerStyle": "caret" | "tilde" | "exact"
  }
}
```

Or just toss something like `defaultSemVerStyle` on the root?

Additionally, since `tilde` isn't an option for `rush add`, I think it would make sense to add it.  This way you can still override the default with a tilde, in case your `rush.json` default is set to, say, caret.

**Comment by Js-Brecht** [comment]
Sorry, mean to tag @octogonz in that last comment

**Comment by octogonz** [comment]
BTW for `@types/` packages, we generally advise an exact version (because these packages do not follow SemVer and frequently have breaking changes in a PATCH update).  As  long as we're improving `rush add`, it would be cool to support that policy as well.

@Js-Brecht as far as the design, I think my leaning would be towards _"Helping developers add a SemVer range that is consistent with the rest of the monorepo"_ (`defaultSemVerStyle`) versus _"Ability to make any CLI parameter have a different default in each repo"_ (`cliDefaults`). :-)

Since this is related to versions, and we're specifying a common policy for all projects, maybe the setting can go in [common-versions.json](https://rushjs.io/pages/configs/common-versions_json/).

Something like:
```js
  // Normally "rush add" will choose the version range that is already in use by other
  // projects in the monorepo.  For a new package that isn't in use, this specifies the
  // default range ("caret", "tilde", "exact")
  "defaultSemVerStyle": "caret",
  // Overrides the "defaultSemVerStyle" for @types packages, which do not follow SemVer
  "defaultSemVerStyleForTypes": "exact",
```

...or if we want to be a bit more flexible, maybe something like this:

```js
  // The first matching rule is used, so more general rules go at the bottom.
  "dependencyVersionDefaults": [
    {
      // Similar to "dependenciesToExclude" from deploy.json, an asterisk
      // matches zero or more characters in the package name.
      "packageNamePattern": "@types/*",
      "semVerStyle": "exact"
    },
    {
      "packageNamePattern": "*",
      "semVerStyle": "tilde"
    }
  ]
```


> Additionally, since `tilde` isn't an option for `rush add`, I think it would make sense to add it. This way you can still override the default with a tilde, in case your `rush.json` default is set to, say, caret.

Agreed. The default isn't "tilde" anyway, it's "Rush decides based on various heuristics."

**Comment by Js-Brecht** [comment]
I like the `dependencyVersionDefaults` idea.  I'll put something together using that.

**Comment by Goues** [comment]
Hey, this has been sitting here for a while and we just found that in our case, using caret range by default would solve a lot of pain for developers. Is there any plan to finalize this?
